### ISSUE TITLE
[java] ClassNamingConventions suggests to add Util for class containing only static constants

### ISSUE THREAD

**Issue by muminc** [pmd/pmd#1164]
**Affects PMD Version:** 

6.4.0

**Rule:**
[ClassNamingConventions](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#classnamingconventions)

**Description:**

A class containg only static constants triggers violation that it should be renamed to end with either Utils or Helper

**Code Sample demonstrating the issue:**

```
public final class StringConstants {
    public static final String YES = "Y";
    public static final String NO = "N";
    private StringConstants(){
    }
}
```

Error Message:

The utility class name 'StringConstants' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'

**Running PMD through:**   Maven 

**Other Info**

You can reproduce this issue in the  unit test by adding this entry

```
<test-code>
        <description>Class containing constants</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
            public final class StringConstants {
                public static final String YES = "Y";
                public static final String NO = "N";
                private StringConstants(){
                }
            }
            ]]></code>
    </test-code>
```

to file :  ClassNamingConventions.xml

**Comment by adangel** [comment]
Hi, thanks for the report! This is actually not a bug, it works as expected.
You can customize the pattern via the property "utilityClassPattern":

```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions">
    <properties>
        <property name="utilityClassPattern" value="[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)" />
    </properties>
</rule>
```

With the pattern value "[A-Z][a-zA-Z0-9]+" you can disable the check for utility classes.

See also https://pmd.github.io/pmd-6.4.0/pmd_rules_java_codestyle.html#classnamingconventions and https://pmd.github.io/pmd-6.4.0/pmd_userdocs_using_properties.html .

**Comment by folder4ben** [comment]
This is actually not a bug if we consider we should check the utilityClassPattern on this class and the question behind could be what is a utility class?

If a utility class is a class where all methods are class methods (static) rather than object methods, should a class with no static method be part of the utility class filter result?

**Comment by oowekyala** [comment]
That depends on the definition of "utility class" you use but you're right to question it. 

>If a utility class is a class where all methods are class methods (static) rather than object methods, 
should a class with no static method be part of the utility class filter result?

That's one valid definition. The definition *we use* is just a class where all *members* are static, not just methods. Choosing one definition or the other seems arbitrary, but I'd argue that our definition makes sense:
* Static methods are static because they don't need a containing object to "make sense". 
* The class in which they're placed only matters for namespacing purposes. Classes with only static methods (your def of utility classes) are just a namespace for static code that is related in some way.
* Similarly, constants make sense outside of the context of an enclosing object, so a class with just constants is just a namespace as well.
* Enforcing a naming convention for utility classes is mainly to make explicit in user code that they're just a "namespace" and shouldn't be used as an object or a type.
* So I think whether a class is constants-only or otherwise, provided all members are static, the object and type declared by the class are useless -which is why the convention should apply.

So I don't see anything wrong with the current behaviour, but that's just my opinion. You could make your case if you don't agree.

I think adding "Constants" to the default patterns would make sense if we keep our definition though

**Comment by AhmadAbdelghany** [comment]
+1 for adding "Constants" to the default patterns.

**Comment by jsotuyod** [comment]
This should be really easy (just check https://github.com/pmd/pmd/issues/1164#issuecomment-393776672 and add proper unit test), so PRs are welcomed to push this forward.
